### DISCUSSION TITLE
Tokenizing Shakesphere

### DISCUSSION THREAD

**Discussion by MartialTerran** [parthasarathydNU/gen-ai-coursework#13] [Q&A]
"Start a Discussion: Read through an implementation and share what you understood or what confused you. Your questions and insights can help everyone learn."

I built the world's smallest functional LLM (GPT-2 architecture) at 1 Megabyte (weights, biases, and embeddings). 
https://huggingface.co/MartialTerran/coherent_text_from_1_megabyte_GPT2_model
This is a tiny working version of this python GPT-2 Transformer model:
https://huggingface.co/MartialTerran/Toy_GPTs_LLMs_for_CPU_Educational/blob/main/Gettysburg_GPT2_v1.4.2.py   employing learned positional embeddings.
Sinusoidal embeddings are optional:
def forward(self, idx, targets=None):
  #positional_embeddings = learned_embeddings
  sinusoidal_embeddings = self.sinusoidal_embedding_table[:, :T, :].to(device) # (T,C)
  positional_embeddings = sinusoidal_embeddings
  x = tok_emb + positional_embeddings # (B,T,C)
# Even without using any positional encoding at all, e.g.,
   # x = tok_emb  (B,T,C)
# this toy model can still memorize and sequence the original Dataset verbatim:

The Gettysburg_GPT2 is configured to train to inference as a "stochastic parrot" repeating the original dataset verbatim.    It uses a custom tokenizer.  In later versions, I found it to be a challenge to use python to tokenized tabs and line-returns expressed as "explicit" text such as "/n" and "/t" while also tokenizing "invisible" line return characters and invisible tabs.  
In your Transformer, 
https://github.com/parthasarathydNU/gen-ai-coursework/blob/main/advanced-llms/generative-pre-trained-transformer.ipynb
the simple vocabulary builder ["chars = list(set(text))"] seems to tokenize "backslash" separately from the "n".  This, I believe, will HINDER the Transformer in training and embeddings.  Because you will have an abundance of extraneous Ns appended to whole words and periodically distributed and preceded by backslash.  Although the model CAN get large enough to learn to handle these line returns, the hyperparameter "n_head = 1" (single head) does not facilitate the transformer learning how to handle the line-returns plus how to distinguish words.  
In my experience, it would be better to add some sophistication to the vocab/tokenizer as follows:  Preprocess the dataset text to merge the sequence "backslashn" as a unicode character,  and encode the sequence "backslashnbackslashn" as a second unicode character.   Then, during tokenization, these two (line return) and (double line return) will be tokenized as DISTINCT TOKENS separate from the adjacent words.  Thus, distinct words will be separated by distinct spaces, punctuation, and by line returns etc.  This will help the model grasp the real relationships between distinct words, punctuation etc.  
Also, because in the tinyshakespeare text dataset, there are line returns within sentences, the dataset should be cleaned to delete most of the single line-returns, because most of them are totally meaningless.  If possible, use/preserve only the line returns following a punctuation such as colon or period or questionmark.  There is no good reason to burden the Transformer with these extraneous linereturn artefacts.  

Also, there should be at least 8 or 16 or 32 attention heads, not n_head = 1, per layer (aka block).

The use of the word "block" (in the hyperparameters) seems to introduce unnecessary ambiguity:
n_block = 4         # Number of Transformer blocks
block_size = 128    # Length of the sequence window

This should be changed to conform:
n_layers = 4         # Number of Transformer layers
max_input_seq = 128    # Length of the input sequence window
max_number_of_positions   # Max Length of the input/output sequence (not including the final predicted token).
max_epochs = 5000    # Total iterations of whole dataset for training
Early_Stop_loss = 0._    # Epoch-Average Loss of whole dataset, for early stop of training (and save checkpoint)

**Comment by parthasarathydNU** [comment]
Thanks for sharing! @MartialTerran . Feel feel to send a PR and I'll have it reviewed.
